### DISCUSSION TITLE
Ztunnel load balancing for L7/HTTP

### DISCUSSION THREAD

**Discussion by karthik-vijayaraju** [istio/istio#52764] [Q&A] [answered]
Hello, I took a look at the new 1.23 release of Istio which has now made Ambient Mesh production ready. 
Considering a HTTP RPC use case and going through the arch I see that traffic interception happens through Ztunnel which routes requests to Waypoint Proxies.

My question is this: If the Waypoint Proxies are scaled out, how does Ztunnel ensure that the new instances get a complement of the requests. My assumption is that it cannot load balance at the HTTP request level, because it operates at L4. So Ztunnel has no control over how many requests are flowing through a connection assuming HTTP keep alive is on. 

But does it have some other way to load balance? Having a connection max-age may be - sure, it will cause retries, but that may be preferable to having a lot of skew in utilisation at Waypoint Proxies and maybe even errors if an older (by age) proxy is overloaded. Thereby rendering scale outs ineffective.

**Comment by keithmattix** [comment]
max-connection-age is what we do for control-plane <-> data plane connections, so it would be reasonable to do between hops with ztunnels and waypoints for all of the reasons you mentioned. Not sure if we do it today though

**Comment by howardjohn** [comment]
Ztunnel does not do L7 load balancing. It behaves the same as kube-proxy in this regard. For request level load balancing you will need a waypoint proxy.

I also disagree that `max-connection-age` is a reasonable solution _in ztunnel_. Its a reasonable solution _in your app_, because your app can gracefully close a connection -- but Ztunnel cannot, it can only terminate a connection abruptly, possibly in the middle of a request.

**Reply by keithmattix** [reply]
How do we ensure newer waypoints get used then? My point wasn't that ztunnel should use max-connection age for all upstream; this would be just for the waypoint

**Reply by howardjohn** [reply, marked as answer]
If we were to do something like this it would be better done at the waypoint layer, where it can send a `connection: close` instead of just a RST

**Reply by keithmattix** [reply]
Sure that's fair, though I'm curious why we don't do that in xDS. Regardless, I imagine that this kind of forced rebalancing is probably a good idea as waypoints get deployed in high throughput environments

**Reply by howardjohn** [reply]
We do that in XDS. The max connection age is set in gRPC which closes the connection at L7, not L4.
